# Data Consistency Checks

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [13]:
# Import Data Sets "products.csv" and "orderswrangled.csv"
df_prods = pd.read_csv(r'C:\Users\Owner\Documents\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [51]:
dfords = pd.read_csv(r'C:\Users\Owner\Documents\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [15]:
# Create a dataframe
df_test = pd.DataFrame()

In [16]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [17]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [18]:
#Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_test[weird]) > 0:
        print(col)


mix


In [19]:
df_test['mix'] = df_test['mix'].astype('str')

In [20]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [21]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [22]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
df_prods.shape

(49693, 5)

In [24]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
# Look for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
df_prods_clean.shape

(49677, 5)

In [29]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
df_prods_clean_no_dups.shape

(49672, 5)

In [31]:
df_prods_clean_no_dups.to_csv(r'C:\Users\Owner\Documents\Instacart Basket Analysis\02 Data\Prepared Data\products_checked.csv')

# Task 3.5

In [32]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Product_id has a minimum of 1 and a max of 49688. The 50% is at 24845, as well as the mean which is expected. However there is a count of 49693 which means we probably have duplicates or null cells.

Aisle_id has a minumum of 1 and a max of 134, The 50% is 69 and the mean is very close at 68 both of which I would expect.

Department_id has a minimum of 1 and a max of 21. 50% is 13 and the mean is 11 both expected values.

Prices has a min of $1.00 and a max of $99,999.00 which seems very and highly unlikely for an instacart item. 50% is at 7 and the mean is 10, indicating we have a large outlier.

In [40]:
# Check for mixed types
for col in dfords.columns.tolist():
    weird = (dfords[[col]].applymap(type) != dfords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (dfords[weird]) > 0:
        print(col)

No columns came back as mixed for orders.csv

In [41]:
#Check for missing values
dfords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
time_order_was_placed          0
days_since_prior_order    206209
dtype: int64

Days since prior order column is missing 206,209 values

In [42]:
dfords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  time_order_was_placed  days_since_prior_order  
count        3.421083e+06           3.421083e+06            3.214874e+06  
mean         2.776219e+00           1.345202e+01            1.111484e+01  
std          2.046829e+00           4.226088e+00            9.206737e+00  
min          0.000000e+00           0.000000e+00            0.000000e+00  
25%          1.000000e+00           1.000000e+01            4.000000e+00  
50%          3.000000e+00           1.300000e+01            7.000000e+00  
75%          5.000000e+00           1.600000e+01            1.500000e+01  
max          6.000000e+00           2.300000e+01            3.000000e+01

Any new customer would not have days since prior order. Since they wouldn't have a prior order. Of the 3,214,874. We have 206,209 values of most likely new customers.

I would make another data set with rows that only contained values and rename it returning customers. 

In [46]:
df_orders_returning_customers = dfords[dfords['days_since_prior_order'].isnull() == False]

In [48]:
df_orders_returning_customers.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
order_number              0
orders_day_of_week        0
time_order_was_placed     0
days_since_prior_order    0
dtype: int64

In [49]:
df_orders_returning_customers.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.214874e+06  3.214874e+06  3.214874e+06  3.214874e+06   
mean   1.710407e+06  1.710675e+06  1.029701e+05  1.819107e+01   
std    9.875882e+05  9.875466e+05  5.953411e+04  1.779950e+01   
min    1.000000e+00  1.000000e+00  1.000000e+00  2.000000e+00   
25%    8.551032e+05  8.555472e+05  5.138500e+04  6.000000e+00   
50%    1.710098e+06  1.710834e+06  1.026620e+05  1.200000e+01   
75%    2.565527e+06  2.565905e+06  1.543720e+05  2.500000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  time_order_was_placed  days_since_prior_order  
count        3.214874e+06           3.214874e+06            3.214874e+06  
mean         2.777637e+00           1.344082e+01            1.111484e+01  
std          2.044923e+00           4.225992e+00            9.206737e+00  
min          0.000000e+00           0.000000e+00            0.000000e+00  
25%          1.000000e+00           1.000000e+01            4.000000e+00  
50%          3.000000e+00           1.300000e+01            7.000000e+00  
75%          5.000000e+00           1.600000e+01            1.500000e+01  
max          6.000000e+00           2.300000e+01            3.000000e+01

In [53]:
dfords_dups = dfords[dfords.duplicated()]

In [54]:
dfords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, time_order_was_placed, days_since_prior_order]
Index: []

I created a new data frame called dfords_dup in order to print back any duplicate rows the program found. It located no records.

In [56]:
df_orders_returning_customers.to_csv(r'C:\Users\Owner\Documents\Instacart Basket Analysis\02 Data\Prepared Data\returningcustomers.csv')